apps/emqx/src/emqx_frame.erl

```diff
@@ -287,14 +287,23 @@ parse_connect(FrameBin, StrictMode) ->
 % Note: return malformed if reserved flag is not 0.
 parse_connect2(
     ProtoName,
-    <<BridgeTag:4, ProtoVer:4, UsernameFlag:1, PasswordFlag:1, WillRetain:1, WillQoS:2, WillFlag:1,
-        CleanStart:1, Reserved:1, KeepAlive:16/big, Rest2/binary>>,
+    <<BridgeTag:4, ProtoVer:4, UsernameFlag:1, PasswordFlag:1, WillRetainB:1, WillQoS:2,
+        WillFlagB:1, CleanStart:1, Reserved:1, KeepAlive:16/big, Rest2/binary>>,
     StrictMode
 ) ->
     case Reserved of
         0 -> ok;
         1 -> ?PARSE_ERR(reserved_connect_flag)
     end,
+    WillFlag = bool(WillFlagB),
+    WillRetain = bool(WillRetainB),
+    case {WillFlag, WillRetain, WillQoS > 0} of
+        %% [MQTT-3.1.2-15]
+        {false, true, _} -> ?PARSE_ERR(invalid_will_retain);
+        %% [MQTT-3.1.2-13]
```

Comment:
```suggestion
        %% [MQTT-3.1.2-11]
```

Reply:
[MQTT-v3.1.1 [MQTT-3.1.2-13]](https://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#:~:text=If%20the%20Will%20Flag%20is%20set%20to%200%2C%20then%20the%20Will%20QoS%20MUST%20be%20set%20to%200%C2%A0(0x00)%20%5BMQTT%2D3.1.2%2D13%5D.)
[MQTT-v5.0 [MQTT-3.1.2-11]](https://docs.oasis-open.org/mqtt/mqtt/v5.0/os/mqtt-v5.0-os.html#:~:text=If%20the%20Will%20Flag%20is%20set%20to%200%2C%20then%20the%20Will%20QoS%20MUST%20be%20set%20to%200%20(0x00)%20%5BMQTT%2D3.1.2%2D11%5D.)

thanks for catching. comments will be updated in later PR.